python/cudf_polars/cudf_polars/dsl/ir.py

```diff
@@ -1867,13 +1881,46 @@ def do_evaluate(
             table = plc.copying.gather(left.table, lg, left_policy)
             result = DataFrame.from_table(table, left.column_names, left.dtypes)
         else:
+            swapped = False
             if how == "Right":
                 # Right join is a left join with the tables swapped
                 left, right = right, left
                 left_on, right_on = right_on, left_on
+                swapped = True
+            if swapped:
+                effective_order = {
+                    "none": "none",
+                    "left": "right",
+                    "right": "left",
+                    "left_right": "right_left",
+                    "right_left": "left_right",
+                }.get(maintain_order, "none")
+            else:
+                effective_order = maintain_order
             lg, rg = join_fn(left_on.table, right_on.table, null_equality)
-            if how == "Left" or how == "Right":
-                # Order of left table is preserved
+            if how in ("Inner", "Left", "Right", "Full") and effective_order != "none":
+                if effective_order in ("left", "left_right"):
+                    lg, rg = cls._reorder_maps(
+                        left.num_rows,
+                        lg,
+                        left_policy,
+                        right.num_rows,
+                        rg,
+                        right_policy,
+                        primary="left",
+                    )
+                elif effective_order in ("right", "right_left"):
+                    lg, rg = cls._reorder_maps(
+                        left.num_rows,
+                        lg,
+                        left_policy,
+                        right.num_rows,
+                        rg,
+                        right_policy,
+                        primary="right",
+                    )
```

Comment:
Once you make this change, maybe it makes sense to just inline `_reorder_maps`. AFAICT it's a single-use function with a long docstring that adds more cognitive load than is really necessary for this one use case.
```suggestion
                lg, rg = cls._reorder_maps(
                    left.num_rows,
                    lg,
                    left_policy,
                    right.num_rows,
                    rg,
                    right_policy,
                    primary="left" if effective_order.startswith("left") else "right"
                )
```

Reply:
I prefer to keep `_reorder_maps`. For me, it makes the logic here easier to follow